### ISSUE TITLE
Report independent status for each job from a Travis-CI

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8730]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8730|
|Type|enhancement|
|Created|2016-08-07 19:21:03Z|
|Branch|https://github.com/twisted/twisted/tree/8730-travis-job-status|

Travis-CI and Appveyour are reporting a single status per build and you can not see what jobs were executed ... and if some of the allow_failure jobs have failed

this tries to workaround this problem by sending custom status for the jobs

<details><summary>Searchable metadata</summary>

```
trac-id__8730 8730
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__8730_travis_job_status 8730-travis-job-status
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__core core
keywords__None None
time__1470597663701147 1470597663701147
changetime__1474534280962927 1474534280962927
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban removed owner|
|-|-|

PR at https://github.com/twisted/twisted/pull/473

The tests are broken on Travis-CI on python3 since txgithub is not yet ported to python3, but before putting more work into this, I would like to see if other people are finding this useful.

With this change a `travis/$TOXENV` status is reported for each job from a Travis-CI build. The status base name is short as the TOXENV can be long.

In this way it should be easier to see that OSX fails or if twistedchecker fails

If people want to have this in Twisted, I can implement it with https://github.com/PyGithub/PyGithub and switch back to txgithub once it is ported on python3

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I absolutely would like to do this with Twisted.  In fact I think it is super important!  The failing tests are kind of a dealbreaker though ;).

The way I'd _prefer_ to see this done would not be by a single travis builder manually reporting status for different build phases, but rather many different travis statuses, like the way Mimic is set up: https://github.com/rackerlabs/mimic/blob/master/.travis.yml#L12-L44 

Mimic also has some other things we should steal, like its requirements.txt pinning and verification: https://github.com/rackerlabs/mimic/blob/master/.travis/install.sh .  I was hoping we could steal most of that stuff and keep it more or less in sync between the two projects.

If you did want to stick with the individual-status-API approach; please do explain why, since it prohibits build parallelism, which makes it impossible for travis to give us more concurrency credits and run our builds faster ;).

As far as the API interface: switching to pygithub is not a big problem, but if the desire is to dogfood with txgithub, then you don't need to use the Python that is being tested to install it; you can make a separate python2 virtualenv that contains it, since it doesn't need to (and in fact should not!) be using the Twisted install being tested, but an upstream release version.  Again though, I don't really see the point of doing it this way since it unnecessarily serializes the build process and delays status creation until the point in the build where the status is reported, rather than travis just manage the whole thing for us.

Although this is very different than the approach I thought we'd take, thanks for working on the problem!  This is an important step forward.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban removed owner|
|-|-|

OK. So I think that Glyph was expecting that this ticket will fix what is now described in [#8737](https://github.com/twisted/twisted/issues/8737) so his review was for another functionality.

As far as I understood from the IRC, Glyph does not really care if the general Travis-CI status is green while the OSX builds are still failing.

I am putting it back for review to get feedback from the actual functionality targeted by this ticket.

If people want to have this functionality, I can fix the txgithub issue :)

To summarise: 

This ticket tried to improve the status visibility for the allow_failure jobs, and also to have a quick report of the actual type of tests that are aggregated by the single Travis-CI status

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_4):

> As far as I understood from the IRC, Glyph does not really care if the general Travis-CI status is green while the OSX builds are still failing.

I do, actually.  I just want it broken out into a separate Travis job.

> This ticket tried to improve the status visibility for the allow_failure jobs, and also to have a quick report of the actual type of tests that are aggregated by the single Travis-CI status

One of the problems I just realized with this technique is that given the way that Travis secrets work, only Twisted organization members will be able to see fine-grained statuses, and there's no real way to fix this.

Consider: if anyone could push, then anyone could just grab the environment variables in their PR, post them somewhere with Requests, and thereby steal the Github auth token for the status app.

However, any non-member contributor can easily add or remove Travis jobs which report independent statuses that we can see on the Travis page.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

I am aware of this Travis-CI secret limitation and this is clearly explained in their documentation.

I hope that in the near future Travis-CI developer will add this functionality.
Before submitting a ticket to Travis-CI asking for this functionality, I wanted to see if other people might find it useful.

The token only has `repo:status` in the scope so maybe we can make it public. I hope that peoples are not that broken to use it for submitting rogue statuses. 

We already have the codecov.io token which is public (https://buildbot.twistedmatrix.com/builders/fedora22-py2.7-coverage/builds/1335/steps/shell_6/logs/stdio) 

---------

Another nice thing about reporting per job status, is that when a travis-ci build failed, you can see what job has failed without having to hit the travis-ci page... and for some tasks this might be enough to give you an idea of what was wrong..

For example once we have a separate job for topfile, you can see it fail and then you can just start fixing you branch without checking travis .... or if you only run the local tests on py2.7 and py3.5 and you see a failure on py3.4 you can then just run the test locally on python3.4

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

Thanks again, adiroiban.

Ultimately I just don't think that this is a worthwhile thing to do.  Almost all other projects report one overall build status.  While it's cool to see "348 of 348 checks passed" or whatever our current pull request indicator is, it also has ugly interactions with github's UI; the little status widget starts scrolling, which is confusing, it's a really long list, and it's easy to miss things.  I prefer to have the one top-level status and the ability to drill down, as clicking on the travis status does.  It's also misleading to report a bunch of statuses which have already been summarized by another status (since travis does have an overall status already).

I should specifically say that while I don't think making the repo:status token public is a good idea, I also am not vetoing this for that reason.  If lots of other users show up saying they'd really prefer to have status checks presented in this way, then I'd be (grudgingly) okay with that concession.  I just think this is a lot of infrastructure glue to give us a status-reporting infrastructure which is perhaps more to some people's taste (yours) than others (mine), but at the same time inconsistent with github's broader vision.

So, with the caveat that others may reopen if they strongly believe that the different presentation of this status-reporting glue is worthwhile enough to justify the maintenance burden, I'm closing.

Thanks again, and sorry for the miscommunication!
